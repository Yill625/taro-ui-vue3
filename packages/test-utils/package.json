{
  "name": "@taro-ui-vue3/test-utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "license": "MIT",
  "dependencies": {
    "vue": "^3.2.0",
    "@vue/test-utils": "^2.0.0-rc.0"
  }
}