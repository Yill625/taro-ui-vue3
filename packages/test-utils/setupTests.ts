import toHaveBeenWarnedInit from './been-warned'

toHaveBeenWarnedInit()