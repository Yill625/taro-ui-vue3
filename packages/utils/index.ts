import propsFactory from './propsFactory'

export * from './common'
export {
  propsFactory
}