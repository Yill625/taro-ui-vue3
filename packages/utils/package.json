{
  "name": "@taro-ui-vue3/utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "license": "MIT",
  "peerDependencies": {
    "@tarojs/taro": "^3.3.0",
    "vue": "^3.2.0"
  },
  "devDependencies": {
    "@vue/test-utils": "^2.0.0-rc.0"
  }
}