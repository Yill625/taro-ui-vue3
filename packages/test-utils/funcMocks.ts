import * as utils from "@taro-ui-vue3/utils/common"
import { pxTransformMockFn, getEnvMockFn } from './@tarojs/taro'

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let pxTransform: jest.SpyInstance

function initMocks() {
  beforeEach(() => {
    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    pxTransform.mockRestore()
  })
}

export default initMocks