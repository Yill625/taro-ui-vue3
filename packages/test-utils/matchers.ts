import * as utils from "@taro-ui-vue3/utils/common"

jest.mock('@taro-ui-vue3/utils/common')

let warn: jest.SpyInstance
let uuid: jest.SpyInstance

const asserted: Set<string> = new Set()
const NOOP = () => { }

// Ensure console.warn is called with expected message
const toHaveBeenWarned: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed = warn.mock.calls.some(args => args[0].indexOf(received) > -1)
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned` +
        (msgs.length
          ? `.\n\nActual messages:\n\n - ${msgs}`
          : ` but no warning was recorded.`)
    }
  }
}

// Ensure previous console.warn is called with expected message
const toHaveBeenWarnedLast: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed =
    warn.mock.calls[warn.mock.calls.length - 1][0].indexOf(received) > -1
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned last.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned last.\n\nActual messages:\n\n - ${msgs}`
    }
  }
}

// Ensure console.warn is called n times with expected message
const toHaveBeenWarnedTimes: jest.CustomMatcher = (received: string, n: number) => {
  asserted.add(received)
  let found = 0
  warn.mock.calls.forEach(args => {
    if (args[0].indexOf(received) > -1) {
      found++
    }
  })

  if (found === n) {
    return {
      pass: true,
      message: () => `expected "${received}" to have been warned ${n} times.`
    }
  } else {
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned ${n} times but got ${found}.`
    }
  }
}

function initWarnMatchers() {
  beforeEach(() => {
    asserted.clear()
    warn = jest.spyOn(console, 'warn').mockImplementation(NOOP)
    uuid = jest.spyOn(utils, 'uuid').mockReturnValue('__tests__2021')

    // NOT working in concurrent mode
    // should avoid use in test.concurrent()
    expect.extend({
      toHaveBeenWarned,
      toHaveBeenWarnedLast,
      toHaveBeenWarnedTimes
    })
  })

  afterEach(() => {
    const assertedArray = Array.from(asserted)
    const nonAssertedWarnings = warn.mock.calls
      .map(args => args[0])
      .filter(received => {
        return !assertedArray.some(assertedMsg => {
          return received.indexOf(assertedMsg) > -1
        })
      })

    uuid.mockRestore()
    warn.mockRestore()

    if (nonAssertedWarnings.length) {
      throw new Error(
        `test case threw unexpected warnings:\n - ${nonAssertedWarnings.join(
          '\n - '
        )}`
      )
    }
  })
}

export default initWarnMatchers