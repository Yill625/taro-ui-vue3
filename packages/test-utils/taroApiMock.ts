import * as utils from "@taro-ui-vue3/utils/common"
import Taro from '@tarojs/taro'
import { createAnimation } from '@tarojs/taro-h5'

import {
  getEnvMockFn,
  pxTransformMockFn,
  chooseImageMockFn,
  querySelectorMockFn
} from './@tarojs/taro'

Taro.pxTransform = pxTransformMockFn
Taro.chooseImage = chooseImageMockFn
Taro.createSelectorQuery = querySelectorMockFn()
Taro.createAnimation = createAnimation

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let getenv: jest.SpyInstance
let pxTransform: jest.SpyInstance
let chooseImage: jest.SpyInstance

function initTaroMockApis() {
  beforeEach(() => {
    jest.mock('@tarojs/taro', () => {
      return Taro
    })

    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    getenv = jest
      .spyOn(Taro, 'getEnv')
      .mockImplementation(getEnvMockFn)

    chooseImage = jest
      .spyOn(Taro, 'chooseImage')
      .mockImplementation(chooseImageMockFn)

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    getenv.mockRestore()
    pxTransform.mockRestore()
    chooseImage.mockRestore()
  })
}

export default initTaroMockApis
