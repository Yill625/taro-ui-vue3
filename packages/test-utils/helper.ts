import { mount, shallowMount, DOMWrapper, VueWrapper } from '@vue/test-utils'
import { VNode, Slot } from '@vue/runtime-core'

export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))

export interface Slots {
  default: Array<string | VNode> | Slot
  [x: string]: Array<string | VNode> | Slot
}

export function mountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return mount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export function genMountFn(
  comp: any,
  components?: Record<string, object>,
  useShallowMount?: boolean
) {
  return (props = {}, slots?: Slots) => {
    return useShallowMount
      ? shallowMountFactory(comp, components, props, slots)
      : mountFactory(comp, components, props, slots)
  }
}

export function shallowMountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return shallowMount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export interface Dimension {
  top?: number
  bottom?: number
  left?: number
  right?: number
  width?: number
  height?: number
}

export function genDelayedSelectorSpy(
  utils: any,
  dimensions: Dimension
): jest.SpyInstance {
  return jest
    .spyOn(utils, 'delayQuerySelector')
    .mockImplementation((_, sel: string, delay?: number) => {
      return new Promise((resolve) => {
        resolve([dimensions])
      })
    })
}

export async function triggerTouchEvents(
  el: DOMWrapper<Element>,
  touchstartDetails: any,
  touchendDetails: any
) {
  const startDetails = {
    touches: [touchstartDetails],
    changedTouches: [touchstartDetails]
  }

  const endDetails = {
    touches: [touchendDetails],
    changedTouches: [touchendDetails]
  }

  await el.trigger('touchstart', startDetails)
  await el.trigger('touchmove', endDetails)
  await el.trigger('touchend', endDetails)
}

export function testPropClassAndStyle(
  mountFn: (props?: {}, slots?: Slots | undefined) => VueWrapper<any>
) {

  it.concurrent('should render prop class', async () => {
    const wrapper = mountFn({ class: 'test' })
    expect(
      wrapper.classes()
    ).toContain('test')
  })

  it.concurrent('should render prop style', async () => {
    const wrapper = mountFn({ style: { color: 'red' } })
    expect(
      wrapper.attributes('style')
    ).toContain('color: red;')
  })
}

export function hexToRGBA(hex: string, opacity = 1) {
  hex = hex.replace('#', '')
  const rgb = hex
    .match(new RegExp('(.{' + hex.length / 3 + '})', 'g'))!
    .map((l) => parseInt(hex.length % 2 ? l + l : l, 16))

  const rgba = rgb.concat(isFinite(opacity) ? opacity : 1)

  return opacity === 1
    ? `rgb(${rgb.join(', ')})`
    : `rgba(${rgba.join(', ')})`
}