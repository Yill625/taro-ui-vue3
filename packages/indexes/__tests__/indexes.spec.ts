import {
  genMountFn,
  genDelayedSelectorSpy
} from '@taro-ui-vue3/test-utils/helper'
import { flushPromises } from '@vue/test-utils'
import { h } from '@vue/runtime-core'
import AtIndexes from '../index.vue'
import * as utils from '@taro-ui-vue3/utils/common'
import { ref } from 'vue'

const mockData = [
  {
    title: 'A', key: 'A', items: [
      { name: ' 阿坝 ', key: 'A' },
      { name: ' 阿拉善 ', key: 'A' }
    ]
  },
  {
    title: 'B', key: 'B', items: [{ name: ' 北京 ', key: 'B' }]
  },
  {
    title: 'J', key: 'J', items: [
      { name: ' 济南 ', key: 'J' },
      { name: ' 佳木斯 ', key: 'J' },
      { name: ' 吉安 ', key: 'J' },
    ]
  }
]

const mountFn = genMountFn(AtIndexes)

describe('AtIndexes', () => {
  it('should render default indexes', async () => {
    const wrapper = mountFn({ isVibrate: false })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it.concurrent('should render an index menu of letters', async () => {
    const wrapper = mountFn({ list: mockData, isVibrate: false })
    const letters = ['Top', 'A', 'B', 'J']
    wrapper.findAll('.at-indexes__menu-item').forEach((el, i) => {
      expect(el.text()).toEqual(letters[i])
    })
  })

  it.concurrent('should render the default slot content', async () => {
    const wrapper = mountFn({ list: mockData, isVibrate: false }, {
      default: [h('view', { class: 'slot' })]
    })
    expect(wrapper.find('#at-indexes__top > .slot').exists()).toBeTruthy()
  })
})

describe('AtIndexes events', () => {
  jest.mock('@taro-ui-vue3/utils/common')
  jest.useFakeTimers()
  let delayedSelector: jest.SpyInstance

  beforeEach(() => {
    delayedSelector = genDelayedSelectorSpy(utils, {
      top: 100,
      height: 60
    })
  })

  afterEach(() => {
    delayedSelector.mockClear()
    delayedSelector.mockRestore()
  })

  it('should render toast and show active letter style when clicking a letter index', async () => {

    const wrapper = mountFn({ list: mockData, isVibrate: false })

    jest.advanceTimersByTime(500) // allow time to calculate all heights
    await flushPromises()
    await wrapper
      .find('.at-indexes__menu-item:nth-child(2)')
      .trigger('tap')
    await wrapper.vm.$nextTick()

    expect(
      wrapper
        .find('.at-indexes__menu-item:nth-child(2)')
        .attributes('style')
    ).toContain('color: white;')

    expect(wrapper.html()).toMatchSnapshot()

    expect(
      wrapper
        .find('.at-toast .toast-body-content__info > text')
        .text()
    ).toEqual('A')
  })

  it('should emit click when clicking on list item', async () => {
    const onClick = jest.fn()
    const wrapper = mountFn({
      list: mockData,
      isVibrate: false,
      onClick
    })

    await wrapper
      .find('#at-indexes__list-B .at-list__item')
      .trigger('tap')

    expect(wrapper.emitted()).toHaveProperty('click')
    expect(onClick).toBeCalled()
  })

  it('should emit scroll-into-view', async () => {
    const jumpToView = ref((k) => { })
    const onScrollIntoView = jest.fn((fn) => {
      jumpToView.value = fn
    })

    const wrapper = mountFn({
      list: mockData,
      isVibrate: false,
      onScrollIntoView
    })

    jest.advanceTimersByTime(500)
    await flushPromises()
    expect(wrapper.emitted()).toHaveProperty('scroll-into-view')
    expect(onScrollIntoView).toBeCalled()

    // jump to J
    jumpToView.value('J')
    jest.advanceTimersByTime(1000) // allow to update state
    await wrapper.vm.$nextTick()

    const jEl = wrapper.find('.at-indexes__menu-item:nth-child(4)')
    expect(jEl.text()).toEqual('J')
    expect(jEl.attributes('style')).toContain('color: white;')

    // jump to Top
    jumpToView.value('top')
    jest.advanceTimersByTime(1000)
    await wrapper.vm.$nextTick()

    const topEl = wrapper.find('.at-indexes__menu-item:nth-child(1)')
    expect(topEl.text()).toEqual('Top')
    expect(topEl.attributes('style')).toContain('color: white;')
  })

  // @TODO: cannot trigger scroll with scroll event details
  it('should render active menu letters when scrolling', async () => {
    const wrapper = mountFn({
      list: mockData,
      isVibrate: false
    })

    jest.advanceTimersByTime(500)
    await flushPromises()

    expect(
      wrapper
        .find('.at-indexes__menu-item:nth-child(1)')
        .attributes('style')
    ).toContain('color: white;')

    await wrapper
      .find('.at-indexes__body')
      .trigger('scroll', {
        detail: { scrollTop: 80 }
      })
    await wrapper.vm.$nextTick()

    // @TODO: handleScroll(e) --> e.detail is 0 ?
    // expect(
    //   wrapper
    //     .find('.at-indexes__menu-item:nth-child(2)')
    //     .attributes('style')
    // ).toContain('color: white;')
  })

  it('should render active letter when touchmoves the menu', async () => {
    const wrapper = mountFn({
      list: mockData,
      isVibrate: false
    })

    jest.advanceTimersByTime(500)
    await flushPromises()

    expect(
      wrapper
        .find('.at-indexes__menu-item:nth-child(1)')
        .attributes('style')
    ).toContain('color: white;')

    await wrapper
      .find('.at-indexes__menu')
      .trigger('touchmove', {
        stopPropagation: jest.fn(),
        preventDefault: jest.fn(),
        touches: [{
          // mocking touchmove to the 3rd menu letter
          // (pageY - startTop) / itemHeight
          // (145 - 100) / (60 / (3 + 1)) = 3
          pageY: 145
        }]
      })
    await wrapper.vm.$nextTick()
    jest.advanceTimersByTime(1000)

    expect(
      wrapper
        .find('.at-indexes__menu-item:nth-child(4)')
        .attributes('style')
    ).toContain('color: white;')
  })
})