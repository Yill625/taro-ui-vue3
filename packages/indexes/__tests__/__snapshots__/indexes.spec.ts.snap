// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtIndexes events should render toast and show active letter style when clicking a letter index 1`] = `
<view class="at-indexes">
  <view class="at-indexes__menu">
    <view class="at-indexes__menu-item" style="">Top</view>
    <view class="at-indexes__menu-item" style="color: white; background-color: rgb(97, 144, 232); border-radius: 40px;">A</view>
    <view class="at-indexes__menu-item">B</view>
    <view class="at-indexes__menu-item">J</view>
  </view>
  <scroll-view class="at-indexes__body" scroll-y="" enable-back-to-top="" scroll-top="60" scroll-into-view="" scroll-with-animation="false">
    <view id="at-indexes__top" class="at-indexes__content"></view>
    <view id="at-indexes__list-A" class="at-indexes__list">
      <view class="at-indexes__list-title">A</view>
      <view class="at-list">
        <view class="at-list__item at-list__item--no-border">
          <view class="at-list__item-container">
            <!--v-if-->
            <!--v-if-->
            <view class="item-content at-list__item-content">
              <view class="item-content__info">
                <view class="item-content__info-title"> 阿坝 </view>
                <!--v-if-->
              </view>
            </view>
            <view class="item-extra at-list__item-extra">
              <!--v-if-->
              <!--v-if-->
              <!--v-if-->
              <!--v-if-->
            </view>
          </view>
        </view>
        <view class="at-list__item at-list__item--no-border">
          <view class="at-list__item-container">
            <!--v-if-->
            <!--v-if-->
            <view class="item-content at-list__item-content">
              <view class="item-content__info">
                <view class="item-content__info-title"> 阿拉善 </view>
                <!--v-if-->
              </view>
            </view>
            <view class="item-extra at-list__item-extra">
              <!--v-if-->
              <!--v-if-->
              <!--v-if-->
              <!--v-if-->
            </view>
          </view>
        </view>
      </view>
    </view>
    <view id="at-indexes__list-B" class="at-indexes__list">
      <view class="at-indexes__list-title">B</view>
      <view class="at-list">
        <view class="at-list__item at-list__item--no-border">
          <view class="at-list__item-container">
            <!--v-if-->
            <!--v-if-->
            <view class="item-content at-list__item-content">
              <view class="item-content__info">
                <view class="item-content__info-title"> 北京 </view>
                <!--v-if-->
              </view>
            </view>
            <view class="item-extra at-list__item-extra">
              <!--v-if-->
              <!--v-if-->
              <!--v-if-->
              <!--v-if-->
            </view>
          </view>
        </view>
      </view>
    </view>
    <view id="at-indexes__list-J" class="at-indexes__list">
      <view class="at-indexes__list-title">J</view>
      <view class="at-list">
        <view class="at-list__item at-list__item--no-border">
          <view class="at-list__item-container">
            <!--v-if-->
            <!--v-if-->
            <view class="item-content at-list__item-content">
              <view class="item-content__info">
                <view class="item-content__info-title"> 济南 </view>
                <!--v-if-->
              </view>
            </view>
            <view class="item-extra at-list__item-extra">
              <!--v-if-->
              <!--v-if-->
              <!--v-if-->
              <!--v-if-->
            </view>
          </view>
        </view>
        <view class="at-list__item at-list__item--no-border">
          <view class="at-list__item-container">
            <!--v-if-->
            <!--v-if-->
            <view class="item-content at-list__item-content">
              <view class="item-content__info">
                <view class="item-content__info-title"> 佳木斯 </view>
                <!--v-if-->
              </view>
            </view>
            <view class="item-extra at-list__item-extra">
              <!--v-if-->
              <!--v-if-->
              <!--v-if-->
              <!--v-if-->
            </view>
          </view>
        </view>
        <view class="at-list__item at-list__item--no-border">
          <view class="at-list__item-container">
            <!--v-if-->
            <!--v-if-->
            <view class="item-content at-list__item-content">
              <view class="item-content__info">
                <view class="item-content__info-title"> 吉安 </view>
                <!--v-if-->
              </view>
            </view>
            <view class="item-extra at-list__item-extra">
              <!--v-if-->
              <!--v-if-->
              <!--v-if-->
              <!--v-if-->
            </view>
          </view>
        </view>
      </view>
    </view>
  </scroll-view>
  <view class="at-toast">
    <!--v-if-->
    <view class="toast-body toast-body--text" style="min-width: 2.1334rem;">
      <view class="toast-body-content">
        <!--v-if-->
        <view class="toast-body-content__info"><text>A</text></view>
      </view>
    </view>
  </view>
</view>
`;

exports[`AtIndexes should render default indexes 1`] = `
<view class="at-indexes">
  <view class="at-indexes__menu">
    <view class="at-indexes__menu-item" style="color: white; background-color: rgb(97, 144, 232); border-radius: 40px;">Top</view>
  </view>
  <scroll-view class="at-indexes__body" scroll-y="" enable-back-to-top="" scroll-top="0" scroll-into-view="" scroll-with-animation="false">
    <view id="at-indexes__top" class="at-indexes__content"></view>
  </scroll-view>
  <!--v-if-->
</view>
`;
